src/crdt.ts

```diff
@@ -151,13 +160,104 @@ export class CRDT<T extends DocTypes> {
 
   // if (snap) await this.clock.applyHead(crdtMeta.head, this.clock.head)
 
-  async allDocs(): Promise<{ result: DocUpdate<T>[]; head: ClockHead }> {
-    await this.ready();
-    const result: DocUpdate<T>[] = [];
-    for await (const entry of getAllEntries<T>(this.blockstore, this.clock.head, this.logger)) {
-      result.push(entry);
+  /**
+   * Retrieve the current set of documents.
+   */
+  allDocs<T extends DocTypes>({ waitFor }: { waitFor?: Promise<unknown> } = {}): QueryResponse<T> {
+    const stream = this.#stream.bind(this);
+
+    return {
+      snapshot: (sinceOpts) => this.#snapshot<T>(sinceOpts, { waitFor }),
+      subscribe: (callback) => this.#subscribe<T>(callback),
+      toArray: (sinceOpts) => arrayFromAsyncIterable(this.#snapshot<T>(sinceOpts, { waitFor })),
+
+      live(opts?: { since?: ClockHead } & ChangesOptions) {
+        return stream<T>({ ...opts, futureOnly: false }, { waitFor });
+      },
+      future() {
+        return stream<T>({ futureOnly: true }, { waitFor });
+      },
+    };
+  }
+
+  #currentDocs<T extends DocTypes>(since?: ClockHead, sinceOptions?: ChangesOptions) {
+    return since ? this.changes<T>(since, sinceOptions) : this.all<T>();
+  }
+
+  #snapshot<T extends DocTypes>(
+    opts: { since?: ClockHead } & ChangesOptions = {},
+    { waitFor }: { waitFor?: Promise<unknown> } = {},
+  ): AsyncGenerator<DocWithId<T>> {
+    const currentDocs = this.#currentDocs.bind(this);
+    const ready = this.ready.bind(this);
+
+    async function* currentDocsWithId() {
+      await waitFor;
+      await ready();
+
+      for await (const doc of currentDocs<T>(opts.since, opts)) {
+        yield docUpdateToDocWithId(doc);
+      }
     }
-    return { result, head: this.clock.head };
+
+    return currentDocsWithId();
+  }
+
+  #subscribe<T extends DocTypes>(callback: (doc: DocWithId<T>) => void) {
+    const unsubscribe = this.clock.onTick((updates: DocUpdate<NonNullable<unknown>>[]) => {
+      updates.forEach((update) => {
+        callback(docUpdateToDocWithId(update as DocUpdate<T>));
+      });
+    });
+
+    return unsubscribe;
+  }
+
+  #stream<T extends DocTypes>(
+    opts: { futureOnly: boolean; since?: ClockHead } & ChangesOptions,
+    { waitFor }: { waitFor?: Promise<unknown> } = {},
+  ) {
+    const currentDocs = this.#currentDocs.bind(this);
+    const ready = this.ready.bind(this);
+    const subscribe = this.#subscribe.bind(this);
+
+    let unsubscribe: undefined | (() => void);
+    let isClosed = false;
+
+    return new ReadableStream<{ doc: DocWithId<T>; marker: QueryStreamMarker }>({
```

Comment:
we need to test that with the worker runtime of Cloudflare. They protect against the direct use of ReadableStream.